extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/exec/ControllerImpl.java

```diff
@@ -1715,12 +1725,95 @@ private void publishSegmentsIfNeeded(
   {
     if (queryKernel.isSuccess() && MSQControllerTask.isIngestion(task.getQuerySpec())) {
       final StageId finalStageId = queryKernel.getStageId(queryDef.getFinalStageDefinition().getStageNumber());
+      queryDef.getFinalStageDefinition().getClusterBy();
 
       //noinspection unchecked
       @SuppressWarnings("unchecked")
       final Set<DataSegment> segments = (Set<DataSegment>) queryKernel.getResultObjectForStage(finalStageId);
+      DataSchema dataSchema = ((SegmentGeneratorFrameProcessorFactory) queryKernel.getStageDefinition(finalStageId)
+                                                                                  .getProcessorFactory()).getDataSchema();
+
+      ShardSpec shardSpec = segments.isEmpty()
+                            ? null
+                            : segments.stream()
+                                                                 .findFirst()
+                                                                 .get()
+                                                                 .getShardSpec();
+      List<String> partitionDimensions = Collections.emptyList();
+
+      if (shardSpec != null && (Objects.equals(shardSpec.getType(), ShardSpec.Type.SINGLE)
+                                || Objects.equals(shardSpec.getType(), ShardSpec.Type.RANGE))) {
+        partitionDimensions = ((DimensionRangeShardSpec) shardSpec).getDimensions();
+      }
+
+      Function<Set<DataSegment>, Set<DataSegment>> compactionStateAnnotateFunction = compactionStateAnnotateFunction(
+          task(),
+          context.jsonMapper(),
+          dataSchema,
+          partitionDimensions
+      );
       log.info("Query [%s] publishing %d segments.", queryDef.getQueryId(), segments.size());
-      publishAllSegments(segments);
+      publishAllSegments(compactionStateAnnotateFunction.apply(segments));
+    }
+  }
+
+  public static Function<Set<DataSegment>, Set<DataSegment>> compactionStateAnnotateFunction(
+      MSQControllerTask task,
+      ObjectMapper jsonMapper,
+      DataSchema dataSchema,
+      List<String> partitionDimensions
+  )
+  {
+    final boolean storeCompactionState = task.getContextValue(
```

Comment:
I'm just thinking if it would be better to move it to query context instead of task context to enable setting it from the web-console. Any thoughts on that?

Reply:
Moved this to query context